### PR TITLE
token-cli: create transfer-fee-basis-points and transfer-fee-maximum-fee flags

### LINKED ISSUE #7020: token-cli: `set-transfer-fee` has different granularity than `create-token`
As reported at https://github.com/anza-xyz/agave/issues/2135:

#### Problem

set-transfer-fee works in whole tokens. create-token with "--transfer-fee" works in 10^(-decimals) tokens.

#### Proposed Solution

Either make them equal or document this more clearly. The way things are now is just asking for errors.

cc @SolBugz

### PR DESCRIPTION
**Pull request by Hrushi20**
Split `transfer-fee` into two separate flags i.e `transfer-fee-basis-points` and `transfer-fee-maximum-fee`. Add a message to deprecate `transfer-fee`.

Fixes #7020

### CODE REVIEW

**Comment by Hrushi20** [comment]
Do I add a test for `transfer-fee-maximum-fee`? Any more specific tests I need to add?

File: token/cli/src/clap_app.rs [hunk 1]
```diff
@@ -763,11 +763,34 @@ pub fn app<'a, 'b>(
                         .value_names(&["FEE_IN_BASIS_POINTS", "MAXIMUM_FEE"])
                         .takes_value(true)
                         .number_of_values(2)
+                        .hidden(true)
```
Comment (joncinque) [inline token/cli/src/clap_app.rs:766]: Let's also make these conflict with the new ones, so that clap will raise the error for us
```suggestion
                        .hidden(true)
                        .conflicts_with("transfer_fee_basis_points")
                        .conflicts_with("transfer_fee_maximum_fee")
```

File: token/cli/src/clap_app.rs [hunk 2]
```diff
@@ -763,11 +763,34 @@ pub fn app<'a, 'b>(
...
+                        ),
+                )
+                .arg(
+                    Arg::with_name("transfer_fee_maximum_fee")
+                        .long("transfer-fee-maximum-fee")
+                        .value_names(&["MAXIMUM_FEE"])
+                        .takes_value(true)
+                        .number_of_values(1)
+                        .help(
+                            "Add a UI amount transfer fee to the mint. \
```
Comment (joncinque) [inline token/cli/src/clap_app.rs:790]: nit:
```suggestion
                            "Add a UI amount maximum transfer fee to the mint. \
```

File: token/cli/src/clap_app.rs [hunk 3]
```diff
@@ -763,11 +763,34 @@ pub fn app<'a, 'b>(
...
+                            The mint authority can set the fee and collected fee"
```
Comment (joncinque) [inline token/cli/src/clap_app.rs:791]: nit:
```suggestion
                            The mint authority can set and collect fees"
```

File: token/cli/src/clap_app.rs [hunk 4]
```diff
@@ -763,11 +763,34 @@ pub fn app<'a, 'b>(
...
+                        .number_of_values(1)
+                        .help(
+                            "Add transfer fee to the mint. \
+                            The mint authority can set the fee.",
+                        ),
+                )
+                .arg(
+                    Arg::with_name("transfer_fee_maximum_fee")
+                        .long("transfer-fee-maximum-fee")
+                        .value_names(&["MAXIMUM_FEE"])
```
Comment (joncinque) [inline token/cli/src/clap_app.rs:786]: we can add a bit more here
```suggestion
                        .value_names(&["MAXIMUM_FEE"])
                        .validator(is_amount)
                        .conflicts_with("transfer_fee")
                        .requires("transfer_fee_basis_points")
```

File: token/cli/src/clap_app.rs [hunk 5]
```diff
@@ -763,11 +763,34 @@ pub fn app<'a, 'b>(
...
+                        .hidden(true)
                         .help(
                             "Add a transfer fee to the mint. \
                             The mint authority can set the fee and withdraw collected fees.",
                         ),
                 )
+                .arg(
+                    Arg::with_name("transfer_fee_basis_points")
+                        .long("transfer-fee-basis-points")
+                        .value_names(&["FEE_IN_BASIS_POINTS"])
```
Comment (joncinque) [inline token/cli/src/clap_app.rs:775]: We can also add a bit more here so that clap catches more issues for us
```suggestion
                        .value_names(&["FEE_IN_BASIS_POINTS"])
                        .conflicts_with("transfer_fee")
                        .requires("transfer_fee_maximum_fee")
```

File: token/cli/src/command.rs [hunk 6]
```diff
@@ -3472,6 +3472,7 @@ pub async fn process_command<'a>(
             let member_address = value_t!(arg_matches, "member_address", Pubkey).ok();
 
             let transfer_fee = arg_matches.values_of("transfer_fee").map(|mut v| {
+                println_display(config,"transfer_fee flag has been deprecated".to_string());
```
Comment (joncinque) [inline token/cli/src/command.rs:3475]: nit: We can give a bit more info
```suggestion
                println_display(config,"transfer-fee has been deprecated and will be removed in a future release. Please specify --transfer-fee-basis-points and --transfer-fee-maximum-fee with a UI amount".to_string());
```

File: token/cli/src/command.rs [hunk 7]
```diff
@@ -3484,6 +3485,45 @@ pub async fn process_command<'a>(
                 )
             });
 
+            let transfer_fee_basis_point = arg_matches.values_of("transfer_fee_basis_points").map(|mut v| {
+                v.next()
+                    .unwrap()
+                    .parse::<u16>()
+                    .unwrap_or_else(print_error_and_exit)
+            });
```
Comment (joncinque) [inline token/cli/src/command.rs:3493]: Let's use `value_of` for simplicity instead
```suggestion
            let transfer_fee_basis_point = value_of::<u16>(arg_matches, "transfer_fee_basis_points");
```

File: token/cli/src/command.rs [hunk 8]
```diff
@@ -3484,6 +3485,45 @@ pub async fn process_command<'a>(
...
+
+            let transfer_fee_maximum_fee = arg_matches.values_of("transfer_fee_maximum_fee").map(|mut v| {
+                v.next()
+                    .unwrap()
+                    .parse::<f64>()
+                    .unwrap_or_else(print_error_and_exit)
+            });
```
Comment (joncinque) [inline token/cli/src/command.rs:3500]: Since we want to map this to a non-ui amount, it's a bit trickier
```suggestion
            let transfer_fee_maximum_fee = value_of::<f64>(
                arg_matches, "transfer_fee_maximum_fee").map(|v| spl_token::ui_amount_to_amount(v, decimals));
```

File: token/cli/src/command.rs [hunk 9]
```diff
@@ -3484,6 +3485,45 @@ pub async fn process_command<'a>(
...
+                    true => Ok(Some((transfer_fee_basis_point.unwrap(), (transfer_fee_maximum_fee.unwrap() * (u64::pow(10, decimals.into())) as f64) as u64))),
+                    false =>  if transfer_fee_basis_point.is_some() {
+                        Err("Missing transfer_fee_maximum_fee flag")
+                    }else if transfer_fee_maximum_fee.is_some() {
+                        Err("Missing transfer_fee_basis_point")
+                    }else{
+                        Ok(None)
+                    }
+                }
+            };
```
Comment (joncinque) [inline token/cli/src/command.rs:3517]: This whole part can be removed thanks to the conflicts logic declared earlier!

File: token/cli/src/command.rs [hunk 10]
```diff
@@ -3484,6 +3485,45 @@ pub async fn process_command<'a>(
...
+
+            let transfer_fee = match transfer_fee {
+                Ok(result) => result,
+                Err(e) => {
+                    eprintln!("{}",e);
+                    exit(1);
+                }
+            };
```
Comment (joncinque) [inline token/cli/src/command.rs:3525]: Here we'll need to combine the transfer fee to make it ready for the call to `command_create_token`, ie

```suggestion
            let transfer_fee = transfer_fee_basis_points.map(|v| (v, transfer_fee_maximum_fee.unwrap()).or(transfer_fee);
```

Maybe you can come up with something better!

File: token/cli/tests/command.rs [hunk 11]
```diff
@@ -123,6 +123,8 @@ async fn main() {
         async_trial!(non_transferable, test_validator, payer),
         async_trial!(default_account_state, test_validator, payer),
         async_trial!(transfer_fee, test_validator, payer),
+        async_trial!(transfer_fee_basis_point, test_validator, payer),
+        // async_trial!(transfer_fee_maximum_fee, test_validator, payer),
```
Comment (joncinque) [inline token/cli/tests/command.rs:127]: nit: Can you remove this comment?

**Review by joncinque — COMMENTED** [review]
Thanks for the contribution, it looks good overall! Mostly some stylistic comments.

Also, can you run `./cargo-nightly fmt --all` from the top of the repo? There's going to be some formatting issues when I run CI

File: token/cli/src/clap_app.rs [hunk 12]
```diff
@@ -763,11 +763,40 @@ pub fn app<'a, 'b>(
...
+                        .help(
+                            "Add transfer fee to the mint. \
+                            The mint authority can set the fee.",
+                        ),
+                )
+                .arg(
+                    Arg::with_name("transfer_fee_maximum_fee")
+                        .long("transfer-fee-maximum-fee")
+                        .value_names(&["MAXIMUM_FEE"])
+                        .takes_value(true)
```
Comment (joncinque) [inline token/cli/src/clap_app.rs:791]: Can you also add a validator for it?
```suggestion
                        .takes_value(true)
                        .validator(is_amount)
```

**Review by joncinque — COMMENTED** [review]
It looks like this is failing some of the clippy lints -- can you run `./cargo-nightly clippy -Zunstable-options --all-targets --features test-sbf  -- --deny=warnings --deny=clippy::arithmetic_side_effects` from the top of the repo and fix whatever's reported?

We're almost there!

**Review by joncinque — APPROVED** [review]
Thanks for your contribution!

thanks
